107-tee-boot-attestation.md

```diff
@@ -0,0 +1,360 @@
+provided by keylime *in addition to* TEE technology. That is, it is reasonable
+to expect that keylime agents will be deployed on CVMs. With that, it is
+reasonable to include TEE attestation mechanisms for environments in which
+keylime will already be running for system integrity monitoring purposes.
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+ * Extend keylime verifier to include support for TEE boot attestation.
+ * Extend TPM measurements to account for TEE boot attestation within keylime.
+ * Fully support confidential computing systems' need for boot *and* runtime
+   attestation within keylime.
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+ * TODO
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+Initially, the TEE handler will work in-tandem with [SVSM], which runs in
+a tenant's firmware and provides privileged operations for CVM guests. SVSM will
+handle TEE attestation before a guest OS boots. SVSM will be responsible for
+unlocking state (provided by the keylime verifier) needed to begin the guest OS
+boot process.
+
+#### Story 1
+ * User launches a confidential VM, pointing it to the URL of the keylime
+   verifier's TEE handler to communicate with for boot attestation.
+ * CVM begins running in SVSM. SVSM gathers all necessary information needed for
+   TEE attestation. Information is dependant on the specific TEE architecture
+   the guest is running on top of (ARM CCA, Intel TDX, AMD SEV-SNP, etc...).
+ * SVSM sends all attestation information to TEE handler.
+ * TEE handler uses information to verify the boot environment.
+ * If boot attestation is successful, TEE handler reads the agent's TEE secret
+   from the registrar DB and sends to CVM. If boot attestation fails, TEE
+   handler sends error message.
+ * SVSM reads response from TEE handler. If successful, SVSM uses the TEE secret
+   to unlock some state and begin booting guest OS, with a keylime agent
+   eventually initializing and integrity monitoring continuing as normal.
+ * SVSM will extend system's TPM state to account for TEE boot attestation.
+   *This is intentially vague at the moment, as the exact design is still under
+   ongoing discussion*.
```

Comment:
This is still under design. Much of the "how to attest" portion is completed, but "what does the verifier return, and how is the vTPM extended from what is returned" is still under ongoing discussion. The ultimate goal is "including" TEE boot attestation within the vTPM measurements.

However, SVSM will complete all of this attestation before Linux is able to boot, as the "secret" in most cases will be a LUKS passphrase to an encrypted disk image for our kernel/OS image. For example, we will begin from SVSM with the SNP/TDX launch measurement, that covers the CVM platform, CVM config, CPU state, and firmware code, where firmware is the combination of SVSM and OVMF.

By validating all of this, we prove integrity of everything that runs prior to executing code from the guest root disk. Assuming that the OVMF binary is built with SecureBoot force enabled, and with suitable certificates, then we know that OVMF will have validated shim, and shim will have validated the UKI kernel. So by attesting the SNP/TDX launch masurement, we've established trust all the way to the start of guest userspace.

Since, we've encrypted the disk image, we can trust that it's what is running after attestation. We can trust the guest userspace, as that's coming out of an encrypted disk image, that we can only unlock if we've passed attestation such that some keys are released.